### ISSUE TITLE
hang when adopting/deleting buckets on ledger close

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/stellar-core#1764]
# Issue type
bug

# Version found
9.2

# Your environment and setup
ubuntu 14.04 LTS

# Issue description
Looks like core's main thread sometimes hangs for a long time when adopting/deleting buckets (when closing a ledger)

## Expected result
It should take a few milliseconds to perform those operations.

## Supporting files

```
2018-08-22T16:20:34.812 GCM6Q [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-31e56dc31107243b/tmp-bucket-75bf9a1ca6e0a3d3.xdr as /data/buckets/bucket-730cffe9873ef504b0d3f646e390b0d1465e413e8fa871729785edb0dca7b169.xdr
2018-08-22T16:20:34.813 GCM6Q [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-31e56dc31107243b/tmp-bucket-8a103cbb60613895.xdr as /data/buckets/bucket-b3a0bbffb820ead7544806beb49f0c9c98e818538da070e34610fbcffc2a4a82.xdr
2018-08-22T16:20:34.813 GCM6Q [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-31e56dc31107243b/tmp-bucket-e55f8059c8bb7097.xdr as /data/buckets/bucket-ffa4c2341f968bd08d64c5a192b12557d35a527675fb4b93838ed736ba52db63.xdr
2018-08-22T16:20:35.579 GCM6Q [Bucket DEBUG] Deleting bucket file /data/buckets/tmp/bucket-31e56dc31107243b/tmp-bucket-d9db4332ce684177.xdr that is redundant with existing bucket
2018-08-22T16:20:41.709 GCM6Q [Bucket DEBUG] Deleting bucket file /data/buckets/tmp/bucket-31e56dc31107243b/tmp-bucket-ded156d60d6f3f03.xdr that is redundant with existing bucket
2018-08-22T16:20:50.130 GCM6Q [Bucket DEBUG] Deleting bucket file /data/buckets/tmp/bucket-31e56dc31107243b/tmp-bucket-ef54cd7f993616f9.xdr that is redundant with existing bucket
2018-08-22T16:20:50.193 GCM6Q [Bucket DEBUG] Adopting bucket file /data/buckets/tmp/bucket-31e56dc31107243b/tmp-bucket-73aaa4bc9f4205cb.xdr as /data/buckets/bucket-41372aa50c7821fd9a0b46a1ddfbb3762333678e60471eeb5f9c76cada21ecb2.xdr
2018-08-22T16:20:50.195 GCM6Q [Ledger DEBUG] Advancing LCL: [seq=19595404, hash=a5490e] -> [seq=19595405, hash=499d0e]

```

**Comment by vogel** [comment]
I think we need more logs for that.

Those `Adopting` and `Deleting` messages come mostly from worker threads after merging is finished. Merging causes `adoptFileAsBucket` calls, which then either moves or deletes files.

I think that delays in logs are result of merging itself, not of moving/renaming files (as this should be fast even on slowest systems), and it is hard to show that without knowledge of history of that instance. For example, we expect merges to be pretty fast on most cases, and when those are slow (as merging top-level buckets), we give plenty of time for them to finish.

So it could be caused by, for example: lot of merges launched at once or few very slow merges. Hard yo say without more logs.

**Comment by MonsieurNicolas** [comment]
There is the ledger number in there, from what I remember it was a node that was running for a while, so starting it up before that time and observing it should be enough.

It makes sense that there seems to be multiple merges happening at the same time (or not? needs to be verified), and we're waiting on the latest merge to complete in order to close the ledger. Could it be that the issue is that the worker pool doesn't leave room for "recent" merges? and/or that merges are done all at once (which doesn't allow to prioritize a new merge like batch add over a lower priority merge like a L9 merge).

**Comment by MonsieurNicolas** [comment]
moving out of 10.2.0 for now, will revisit if we see this happen again

**Comment by MonsieurNicolas** [comment]
Got some more information from some machines:
```
2019-02-19T16:58:44.774 GCGB2 [Herder INFO] Quorum information for 22560062 : {"agree":3,"delayed":0,"disagree":0,"fail_at":2,"hash":"75b44e","missing":0,"phase":"EXTER
NALIZE","validated":true}

2019-02-19T16:58:44.778 GCGB2 [Ledger INFO] Got consensus: [seq=22560064, prev=333859, tx_count=11, sv: [  txH: 640674, ct: 1550595524, upgrades: [ ] ]]
2019-02-19T16:58:44.781 GCGB2 [Tx INFO] applying ledger 22560064 (txs:11, ops:49)
2019-02-19T16:58:48.218 GCGB2 [default INFO] 'FutureBucket: merge' executed after 8.336 s
2019-02-19T16:59:02.142 GCGB2 [default INFO] 'Delete redundant bucket' hung for 2.319 s
2019-02-19T16:59:02.142 GCGB2 [default INFO] 'FutureBucket: merge' executed after 17.287 s
2019-02-19T16:59:02.157 GCGB2 [default INFO] 'FutureBucket: merge' executed after 17.302 s
2019-02-19T16:59:02.162 GCGB2 [default INFO] 'FutureBucket: merge' executed after 17.307 s
2019-02-19T16:59:02.164 GCGB2 [default INFO] 'FutureBucket: merge' executed after 17.302 s
2019-02-19T16:59:02.165 GCGB2 [default INFO] 'Resolving bucket' hung for 17.303 s
2019-02-19T16:59:02.191 GCGB2 [History INFO] Publishing 1 queued checkpoints [22560063-22560063]: Retrying in 0 sec: prepare-snapshot
2019-02-19T16:59:02.193 GCGB2 [Ledger INFO] Closed ledger: [seq=22560064, hash=8da99d]
```

Issue might be that we don't distinguish merges required to actually close a ledger from merges that can run in the background.

One thing I am wondering: a simple fix might be to just spin up 10+extra worker threads, enough to deal with "wost case" merges and more (with lower priority than the main thread to avoid too much trashing) instead of using num cores?

@graydon what do you think?

**Comment by graydon** [comment]
@MonsieurNicolas Certainly possible that throwing more threads at it would help, and if this is causing problems in production systems that's not a bad patch to try; I'm only slightly worried that it would make things worse, and/or burden the systems needlessly, if we start having (or are already having) oversubscription of IO resources.

A more-involved patch would involve two separate pieces:

  - Break long-running background work up so it bounces off the background queue / reschedules itself with reasonable frequency.
  - Bring in support for [asio's priority-queue executor](https://github.com/chriskohlhoff/asio/blob/master/asio/src/examples/cpp14/executors/priority_scheduler.cpp) so we can prioritize which background tasks should be preempted by which other ones.

This is probably just overengineering / further-reimplementing threads ourselves, on the mere speculation that the IO subsystem might get cranky when 4 or 5 merges are running concurrently on a 2-way machine. "More threads" is probably adequate, and certainly easier to engineer: one line patch!

**Comment by MonsieurNicolas** [comment]
Cool. @vogel can you look into doing it with more threads then? Easiest is probably to add a new config for how many worker threads we should have, with a a good default.
On the default number: I would start with 10.
Ideally we can set those to have a slightly lower priority (ie: "niced") than the main thread to avoid impacting the main thread, I don't know if asio allows to do this.

**Comment by vogel** [comment]
Lets try!
